clients/src/main/java/org/apache/kafka/common/protocol/Errors.java

```diff
@@ -109,531 +109,98 @@
-        }
-    }),
-    GROUP_ID_NOT_FOUND(69, "The group id does not exist", new ApiExceptionBuilder() {
-        @Override
-        public ApiException build(String message) {
-            return new GroupIdNotFoundException(message);
-        }
-    }),
-    FETCH_SESSION_ID_NOT_FOUND(70, "The fetch session ID was not found",
-        new ApiExceptionBuilder() {
-            @Override
-            public ApiException build(String message) {
-                return new FetchSessionIdNotFoundException(message);
-            }
-    }),
-    INVALID_FETCH_SESSION_EPOCH(71, "The fetch session epoch is invalid",
-        new ApiExceptionBuilder() {
-            @Override
-            public ApiException build(String message) {
-                return new InvalidFetchSessionEpochException(message);
-            }
-    }),
-    LISTENER_NOT_FOUND(72, "There is no listener on the leader broker that matches the listener on which metadata request was processed",
-        new ApiExceptionBuilder() {
-            @Override
-            public ApiException build(String message) {
-                return new ListenerNotFoundException(message);
-            }
-    }),;
+            "return this exception.", UnknownProducerIdException::new),
+    REASSIGNMENT_IN_PROGRESS(60, "A partition reassignment is in progress", ReassignmentInProgressException::new),
+    DELEGATION_TOKEN_AUTH_DISABLED(61, "Delegation Token feature is not enabled.", DelegationTokenDisabledException::new),
+    DELEGATION_TOKEN_NOT_FOUND(62, "Delegation Token is not found on server.", DelegationTokenNotFoundException::new),
+    DELEGATION_TOKEN_OWNER_MISMATCH(63, "Specified Principal is not valid Owner/Renewer.", DelegationTokenOwnerMismatchException::new),
+    DELEGATION_TOKEN_REQUEST_NOT_ALLOWED(64, "Delegation Token requests are not allowed on PLAINTEXT/1-way SSL channels and " + "on delegation token authenticated channels.", UnsupportedByAuthenticationException::new),
+    DELEGATION_TOKEN_AUTHORIZATION_FAILED(65, "Delegation Token authorization failed.", DelegationTokenAuthorizationException::new),
+    DELEGATION_TOKEN_EXPIRED(66, "Delegation Token is expired.", DelegationTokenExpiredException::new),
+    INVALID_PRINCIPAL_TYPE(67, "Supplied principalType is not supported", InvalidPrincipalTypeException::new),
+    NON_EMPTY_GROUP(68, "The group is not empty", GroupNotEmptyException::new),
+    GROUP_ID_NOT_FOUND(69, "The group id does not exist", GroupIdNotFoundException::new),
+    FETCH_SESSION_ID_NOT_FOUND(70, "The fetch session ID was not found", FetchSessionIdNotFoundException::new),
+    INVALID_FETCH_SESSION_EPOCH(71, "The fetch session epoch is invalid", InvalidFetchSessionEpochException::new),
+    LISTENER_NOT_FOUND(72, "There is no listener on the leader broker that matches the listener on which metadata request was processed", ListenerNotFoundException::new),;
 
+    @FunctionalInterface
     private interface ApiExceptionBuilder {
```

Comment:
Do we need this at all or can we use one of the Java 8 functional interfaces?

Reply:
Hmm, thinking of it, we could use `java.util.Function<String, ApiException>`. Let me update this PR to see how it looks.